### DISCUSSION TITLE
Multiple insert issue on event

### DISCUSSION THREAD

**Discussion by shashikantpawar21** [Eventuous/eventuous#99] [Q&A]
Hi Alexey, 
Not sure why the previous conversation is locked... Am i asking too many question.. Sorry for the same 
 ..I am trying to solve most problem though have issue with mongo projection updating 

For multiple insert, I did tried CollectionOperation but it looks for bool return type.. 

I did tried updating mongo table by updating the table but it does not update the position column so seems checkpoint will not work in this case

**Comment by alexeyzimarev** [comment]
The conversation is locked as its topic was a welcome message, which wasn't intended for any replies.

MongoDB doesn't have tables, it has documents and fields in the document, sorry for being picky on the terminology.

The position field is for you to use. If you don't use it - I don't care. The checkpoint is stored in the checkpoints collection.

You are right that it's currently impossible to build a collection operation that would also set the stream position in the updated document. It's a missing feature, I added an issue https://github.com/Eventuous/eventuous/issues/100

If you explain with more details what you are trying to do, I might have another solution. For example, you can write "I handle this event, and I update or insert multiple documents, where document fields get set from event properties like that", add some code that you have, etc.

**Reply by alexeyzimarev** [reply]
Ok, I was wrong. You get the context when building a collection operation. I am wondering if the current code that gets the position from context items is actually correct. There's a field in the context for that, I will check.

```caharp
        On<V1.BookingCancelled>(
            context => new ValueTask<Operation<MyBookings>>(
                new CollectionOperation<MyBookings>(
                    (collection, token) =>
                        collection.UpdateManyAsync(
                            x => x.Bookings.Select(b => b.BookingId).Contains(context.Message.BookingId),
                            Builders<MyBookings>.Update
                                .PullFilter(
                                    x => x.Bookings,
                                    x => x.BookingId == context.Message.BookingId
                                )
                                .Set(x => x.Position, context.GlobalPosition),
                            cancellationToken: token
                        )
                )
            )
        );
```

**Reply by alexeyzimarev** [reply]
I think this might make your life a bit easier. I just added `InsertMany`, took me 5 min.
https://github.com/Eventuous/eventuous/issues/101

**Reply by shashikantpawar21** [reply]
Thank you for quick help ... I see it is merged and I will do check it. Sorry, I am bit confused with release version .. 
I hope after upgrading to **0.8.4** version of eventuous. I will see the InsertMany Api?

**Reply by alexeyzimarev** [reply]
It hasn't been released as a stable package, but you can try it by installing the latest preview package from myget.

**Reply by shashikantpawar21** [reply]
issue is still not resolved, to put issue in a better way, from the example you shared in your previous message.

I am trying to update a field of MyBooking.Bookings.**CheckInDate**  . how to do this as with mongo abstractions method?

**Reply by alexeyzimarev** [reply]
I am not sure what issue are you talking about, and what updating a booking check-in date has to do with multiple inserts. When you update the MyBooking document, you use the guest id and specify the MongoDB native update using the update builder. Eventuous _does not_ have an abstraction for building filters or updates, you can refer to the MongoDB .NET client documentation for that.

With regards to the new API for projectors, here's an example:

```csharp
        On<V1.BookingCancelled>(
            b => b.UpdateOne
                .Filter(
                    (evt, doc)
                        => doc.Bookings.Select(booking => booking.BookingId).Contains(evt.BookingId)
                )
                .Update(
                    (evt, update) =>
                        update.PullFilter(
                            x => x.Bookings,
                            x => x.BookingId == evt.BookingId
                        )
                )
        );
```

**Reply by shashikantpawar21** [reply]
This does not work or may i am lacking to understand filterbuilder and updateDefinition.. 
So far only facing difficulties in handling this mongo projection handler 

Is there a way to try something like below 

 On<Event>(evt => Project(evt));

And we can do anything inside Project(evt) method with evt .. calling database and using normal mongo queries to do the job ?

**Reply by shashikantpawar21** [reply]
Also, With the current handlers, it becomes difficult to get the record from mongo document if we dont know the id field

**Reply by alexeyzimarev** [reply]
Honestly, I don't know what you mean. You don't provide any code or context for your questions, so I am having a hard time understanding what the issue is. The `On<T>` method accepts the `ProjectTypedEvent<T, TEvent>` where `T` is the document type. It allows you to do anything if you don't want to use projector builders. Everything else is just a bunch of helpers.

**Reply by alexeyzimarev** [reply]
I also mentioned many times before that Eventuous provides helpers for building operations, but it uses the MongoDB C# driver elements like filter and update builders. If you don't know how they work, you need to refer to MongoDB C# driver documentation.
